Plane: Landing abort using rc input throttle >95%

magicrub opened the pull request:
- Abort auto-landing when RC throttle input > 95%. 
- Uses new param LAND_ABORT_THR default is 0 (disabled)

I'm not happy with the name of this param, I'm more than happy to change it if someone has a better suggestion.

To work properly this depends on https://github.com/diydrones/ardupilot/pull/2269 but it is otherwise inert without it.

iskess commented:
Very happy to see so much progress on this!
I think the parameter name is a good one. 
One tiny observation, the description says 90% rather than 95%
// @Description: Allow a landing abort to trigger with a throttle > 90%

jinchengde commented:
I think there may some question for abort mission
1. if the plane is in low altitude, and if the mission have auto-take-off Item, then we'd better execute the take-off first to arrive the safe altitude, if we execute the restart landing directly, it may cause cash.
2. Is there any confuse for the throttle nudge?

magicrub commented:
Ahh, thanks! I flew on 90 but changed it to 95 when I submitted the PR,
felt like a safer number. I'll update it.
On Aug 25, 2015 7:20 PM, "Iam Bouret" <notifications@github.com> wrote:

> Very happy to see so much progress on this!
> I think the parameter name is a good one.
> One tiny observation, the description says 90% rather than 95%
> // @Description <https://github.com/Description>: Allow a landing abort
> to trigger with a throttle > 90%
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diydrones/ardupilot/pull/2759#issuecomment-134794812>.
>

magicrub commented:
@jinchengde Right now this one does not interact with take-off at all. This is just a trigger to initiate an abort landing.
@tridge I'm not sure if we should put this in now or wait until we add the abort flight stage to address the low-altitude anti-stall stuff

iskess commented:
@jinchengde,
To answer your other question, the throttle nudge is not useful during landing approach since the aircraft should not be targeted faster than trim speed. During this phase of flight, the throttle lever will be used to trigger an aborted landing. The user will have to train themselves to not leave the throttle up, or it will command a nudge to max speed after the abort is completed and the plane proceeds to its next WP. 

Thinking about this more, it is possible that an airplane with a wide speed range  is using TECS_LAND_ARSPD to fly at an approach speed much closer to stall (1.3 Vso) . In the case of wind shear on approach, the operator might want to nudge the throttle to increase speed to protect the stall margin. However, I don't think it should be possible to nudge past TRIM_ARSPD_CM. So it should be designed that during approach, the bottom half of the throttle stick range become active for nudge, rather than the top half in other phases of flight. This still creates no conflict with using 95% throttle as an abort trigger. This also gives the operator the flexibility to delay decrease in target speed until on short final, like after crossing the tree line for example. 
If you guys like this idea, perhaps I should open a separate issue for it? @Tridge, I would think this would be helpful for the larger planes you fly with their wide speed envelopes.

magicrub commented:
Thanks Iam,

Yes, please open a new issue for that. It's unrelated to abort and more
about safe landing.

Onc thing that I think we need to test is if a flaps come in to play at all
on the abort. Normally they would help but I just want to make sure they
reset correctly.

On Thu, Aug 27, 2015 at 1:19 PM, Iam Bouret <notifications@github.com>
wrote:

> @jinchengde <https://github.com/jinchengde>,
> To answer your other question, the throttle nudge is not useful during
> landing approach since the aircraft should not be targeted faster than trim
> speed. During this phase of flight, the throttle lever will be used to
> trigger an aborted landing. The user will have to train themselves to not
> leave the throttle up, or it will command a nudge to max speed after the
> abort is completed and the plane proceeds to its next WP.
>
> Thinking about this more, it is possible that an airplane with a wide
> speed range is using TECS_LAND_ARSPD to fly at an approach speed much
> closer to stall (1.3 Vso) . In the case of wind shear on approach, the
> operator might want to nudge the throttle to increase speed to protect the
> stall margin. However, I don't think it should be possible to nudge past
> TRIM_ARSPD_CM. So it should be designed that during approach, the bottom
> half of the throttle stick range become active for nudge, rather than the
> top half in other phases of flight. This still creates no conflict with
> using 95% throttle as an abort trigger. This also gives the operator the
> flexibility to delay decrease in target speed until on short final, like
> after crossing the tree line for example.
> If you guys like this idea, perhaps I should open a separate issue for it?
> @Tridge <https://github.com/Tridge>, I would think this would be helpful
> for the larger planes you fly with their wide speed envelopes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diydrones/ardupilot/pull/2759#issuecomment-135541357>.
>

magicrub commented:
alright, I've done a whole makeover on this PR. I've converted it to use a new 5th flight stage called FLIGHT_LAND_ABORT.

- enabled via new param LAND_ABORT_THR default is 0 (disabled)
- Triggered via 95% throttle during landing
- This utilizes all takeoff params for right now (TKOFF_THR), we can make this better later where needed
- added mission item command to NAV_LAND which is the abort takeoff altitude. If 0 then use last used takeoff alt if available, else use 50m
- hold heading, just like takeoff, until altitude is reached.
- pitch is constrained like takeoff pitch, or else 10deg if not available
- After abort altitude is reached, the normal landing restart happens (DO_LAND_START or decrement mission)

iskess commented:
This is great! 
Just to clarify what you mean by "added mission item command to NAV_LAND"...

Two of the "|Empty|" fields in the summary below will now become "abort altitude" and "abort pitch"?

///<summary> Land at location |Empty| Empty| Empty| Desired yaw angle.| Latitude| Longitude|
Altitude| </summary>
LAND=21,

magicrub commented:
It's altitude only for this PR.
On Aug 29, 2015 1:29 AM, "Iam Bouret" <notifications@github.com> wrote:

> This is great!
> Just to clarify what you mean by "added mission item command to
> NAV_LAND"...
>
> Two of the "|Empty|" fields in the summary below will now become "abort
> altitude" and "abort pitch"?
>
> ///
> Land at location |Empty| Empty| Empty| Desired yaw angle.| Latitude|
> Longitude|
> Altitude|
> LAND=21,
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diydrones/ardupilot/pull/2759#issuecomment-135961611>.
>

magicrub commented:
Sorry, I mean the extra mission option is for abort altitude for it to
climb to, pitch and throttle settings are from your previous takeoff. The
land altitude is the same behavior.
On Aug 29, 2015 10:42 AM, tom@airphrame.com wrote:

> It's altitude only for this PR.
> On Aug 29, 2015 1:29 AM, "Iam Bouret" <notifications@github.com> wrote:
>
>> This is great!
>> Just to clarify what you mean by "added mission item command to
>> NAV_LAND"...
>>
>> Two of the "|Empty|" fields in the summary below will now become "abort
>> altitude" and "abort pitch"?
>>
>> ///
>> Land at location |Empty| Empty| Empty| Desired yaw angle.| Latitude|
>> Longitude|
>> Altitude|
>> LAND=21,
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/diydrones/ardupilot/pull/2759#issuecomment-135961611>
>> .
>>
>

magicrub commented:
Rebased on master because https://github.com/diydrones/ardupilot/pull/2269 was merged

magicrub commented:
I've also added a PR https://github.com/mavlink/mavlink/pull/435 for the mission documentation

magicrub commented:
flight tested, works great

tridge commented on ArduPlane/Parameters.h:
```diff
@@ -135,6 +135,7 @@ class Parameters {
         k_param_trim_rc_at_start,
         k_param_hil_mode,
         k_param_land_disarm_delay,
+        k_param_land_abort_throttle_enable,
```
this changes the enum values of several parameters which breaks existing configs
if you have flown this you may have a corrupt eeprom

tridge commented on ArduPlane/commands_logic.cpp:
```diff
@@ -341,6 +341,21 @@ void Plane::do_land(const AP_Mission::Mission_Command& cmd)
+    if (cmd.p1 > 0) {
+        auto_state.takeoff_altitude_rel_cm = (int16_t)cmd.p1 * 100;
+    } else if (auto_state.takeoff_altitude_rel_cm <= 0) {
+        auto_state.takeoff_altitude_rel_cm = 5000;
```
I think a default of 30m would be better. At 10 degrees a lot of models are out of easy control range by the time they get to 50m

tridge commented on ArduPlane/landing.cpp:
```diff
@@ -25,6 +25,26 @@ bool Plane::verify_land()
+            steer_state.hold_course_cd = -1;
+            next_WP_loc = current_loc;
+            mission.stop();
+            restart_landing_sequence();
```
if the restart of the landing sequence fails won't we end up flying off forever?
I think if it fails we probably need to RTL. Either that or we work out what we are going to do before we start the abort, and don't abort the landing if we don't know for sure we can restart the landing sequence

magicrub replied:
The only way this would fail is if you've never flown to another waypoint with a lat/lng. For example, if you did TAKEOFF then LAND then it would not abort but a mission like that doesn't make sense. However, I agree that having a fly-away is horrible and should never be possible. Sure, I can see a fall-back being RTL. Wouldn't hurt. I'll put one in.

magicrub commented:
@tridge Thank you for supplying feedback. I applied your suggestions which I all agreed with. Is there anything else holding up this PR? Would be nice to get into the beta

tridge commented:
the only problem I can see now is the setting of hold_course_cd in do_land().
I think this means that if you do a landing and then immediately reset the mission and do a takeoff that the takeoff will ignore the direction the plane is pointing at the start of the takeoff and will instead try to hold course along the previously used landing course. This is because verify_takeoff() uses hold_course_cd to determine if it needs to set the takeoff course.
The hold_course_cd variable is really quite bad. I think we're going to need to split it into multiple variables, with a separate one for takeoff and land at least.

magicrub commented:
After talking about it on mumble, we decided sharing the takeoff
hold_course_cd variable
is probably a bad idea and just so it won't conflict it would be better to
just make a new variable. While implementing that I realized that this PR
also implements the mode-switch abort feature, which is different, so that
means we will *always* have a prev_wp and next_wp and therefore can always
calculate the bearing on-demand during the abort, it does not need to be
pre-calculated and stored.

I pushed that change and a rebase -i.

-TomP

On Tue, Sep 8, 2015 at 12:29 AM, Andrew Tridgell <notifications@github.com>
wrote:

> the only problem I can see now is the setting of hold_course_cd in
> do_land().
> I think this means that if you do a landing and then immediately reset the
> mission and do a takeoff that the takeoff will ignore the direction the
> plane is pointing at the start of the takeoff and will instead try to hold
> course along the previously used landing course. This is because
> verify_takeoff() uses hold_course_cd to determine if it needs to set the
> takeoff course.
> The hold_course_cd variable is really quite bad. I think we're going to
> need to split it into multiple variables, with a separate one for takeoff
> and land at least.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diydrones/ardupilot/pull/2759#issuecomment-138460602>.
>

magicrub commented:
flew this today, tested with and without a DO_LAND_START in the mission.

tridge commented:
applied, with a small fix to use @Values for the new parameter
thanks Tom!

iskess commented:
Landing Abort took off in an unexpected direction, about 90 degrees off track. 
![abort heading](https://cloud.githubusercontent.com/assets/4913946/13377990/d9283436-dd97-11e5-845f-9fed7a240918.JPG)
[2016-02-25 13-31-52.zip](https://github.com/diydrones/ardupilot/files/149783/2016-02-25.13-31-52.zip)
